### PR TITLE
refactor dsymbol.d: add some inout and const

### PR DESCRIPTION
**Pull request by WalterBright**
These are called from everywhere, if they can't be called with const/mutable arguments, then pretty much nothing can be const.

### CODE REVIEW

File: src/dsymbol.d [hunk 1]
```diff
@@ -460,20 +460,20 @@ extern (C++) class Dsymbol : RootObject
     // instantiated speculatively, eg from inside is(typeof()).
     // Return the speculative template instance it is part of,
     // or NULL if not speculative.
-    final TemplateInstance isSpeculative()
+    final inout(TemplateInstance) isSpeculative() inout
     {
-        Dsymbol par = parent;
+        Dsymbol par = cast()parent;
```
Comment (ZombineDev) [inline src/dsymbol.d]: Would it be possible to remove the casts if we make the `par` and `ti` variable declarations `auto`?
Or does `isTemplateInstance` and `toParent` require mutable `this`?

**Comment by ZombineDev** [comment]
LGTM. I'm just wondering if the the casts needed in a couple of places could be removed if we used `auto`, feel free to merge if it's not possible or it requires more extensive changes.

File: src/dsymbol.d (hunk 1 above)
Comment (WalterBright) [inline src/dsymbol.d]: No, and no. The trouble is that `inout` means a variable cannot be assigned to.

**Comment by yebblies** [comment]
Headers.  These could use recursion instead of casts...

**Comment by WalterBright** [comment]
> Headers.

C++ doesn't support inout.

> recursion

is slow.

**Comment by yebblies** [comment]
> C++ doesn't support inout.

So make them const.

**Comment by WalterBright** [comment]
> So make them const.

That isn't how const works - an inout function can accept/return mutable arguments as well. It's best to just leave them be.

File: src/dsymbol.d (hunk 1 above)
Comment (ZombineDev) [inline src/dsymbol.d]: Sorry, I don't know how I managed to overlook this.

Perhaps Rebindable could be used to encapsulate tail-only const and to make the code `@safe`. Anyway it looks good for now.

File: src/dsymbol.d [hunk 2]
```diff
@@ -392,13 +392,13 @@ extern (C++) class Dsymbol : RootObject
...
+    final inout(Dsymbol) pastMixin() inout
     {
-        Dsymbol s = this;
+        Dsymbol s = cast()this;
         //printf("Dsymbol::pastMixin() %s\n", toChars());
         while (s && s.isTemplateMixin())
             s = s.parent;
-        return s;
+        return cast(inout(Dsymbol))s;
     }
```
Comment (andralex) [inline src/dsymbol.d]: Time to eat our dogfood. TDPL says black on white we do tail call optimization. A variety of public online resources advertise the same. This function must be:

```d
final inout(Dsymbol) pastMixin() inout
{
    if (!isTemplateMixin) return this;
    if (!parent) return null;
    return parent.pastMixin;
}
```

Thanks.

File: src/dsymbol.d [hunk 3]
```diff
@@ -431,27 +431,27 @@ extern (C++) class Dsymbol : RootObject
...
     /// ditto
-    final Dsymbol toParent2()
+    final inout(Dsymbol) toParent2() inout
     {
-        Dsymbol s = parent;
+        Dsymbol s = cast()parent;
         while (s && s.isTemplateInstance())
             s = s.parent;
-        return s;
+        return cast(inout(Dsymbol))s;
```
Comment (andralex) [inline src/dsymbol.d]:
```d
    final inout(Dsymbol) toParent2() inout
    {
        if (!parent) return null;
        if (!!s.isTemplateInstance) return this;
        return parent.toParent2;
    }
```

File: src/dsymbol.d [hunk 4]
```diff
@@ -431,27 +431,27 @@ extern (C++) class Dsymbol : RootObject
...
-    final TemplateInstance isInstantiated()
+    final inout(TemplateInstance) isInstantiated() inout
     {
-        for (Dsymbol s = parent; s; s = s.parent)
+        for (Dsymbol s = cast()parent; s; s = s.parent)
         {
             TemplateInstance ti = s.isTemplateInstance();
             if (ti && !ti.isTemplateMixin())
-                return ti;
+                return cast(inout(TemplateInstance))ti;
```
Comment (andralex) [inline src/dsymbol.d]:
```d
    final inout(TemplateInstance) isInstantiated() inout
    {
        if (!parent) return null;
        auto ti = parent.isTemplateInstance;
        if (ti && !ti.isTemplateMixin) return ti;
        return parent.isInstantiated;
    }
```

File: src/dsymbol.d [hunk 5]
```diff
@@ -460,20 +460,20 @@ extern (C++) class Dsymbol : RootObject
...
         while (par)
         {
             TemplateInstance ti = par.isTemplateInstance();
             if (ti && ti.gagged)
-                return ti;
+                return cast(inout(TemplateInstance))ti;
             par = par.toParent();
         }
         return null;
     }
```
Comment (andralex) [inline src/dsymbol.d]:
```d
    final inout(TemplateInstance) isSpeculative() inout
    {
        if (!parent) return null;
        auto ti = parent.isTemplateInstance;
        if (ti && ti.gagged) return ti;
        Dsymbol par = cast()parent;
        auto gramps = par.toParent;
        return gramps ? gramps.isSpeculative : null;
    }
```

I couldn't avoid the two tests, but I suspect things can be refactored better.

**Comment by andralex** [comment]
Overall: we should use the styles we recommend.

File: src/dsymbol.d (hunk 5 above)
Comment (WalterBright) [inline src/dsymbol.d]: This one wouldn't tail recurse until I split up the `?:` into an `if`.

**Comment by WalterBright** [comment]
Replaced with Andrei's versions, except had to adjust the ?: on the last one.

**Comment by ZombineDev** [comment]
LGTM

File: src/dsymbol.d [hunk 6]
```diff
@@ -431,49 +432,47 @@ extern (C++) class Dsymbol : RootObject
...
-        return null;
+        if (!parent)
+            return null;
+        auto ti = parent.isTemplateInstance();
+        if (ti && ti.gagged)
+            return ti;
+        auto gramps = parent.toParent();
+        if (!gramps)
+            return null;
+        return parent.isSpeculative();
```
Comment (andralex) [inline src/dsymbol.d]: Wouldn't this be `gramps.isSpeculative`?

Comment (WalterBright) [inline src/dsymbol.d]: It was, but failed the test suite. It's subtle, but parent. is correct. Read the before and after code carefully :-)

Comment (andralex) [inline src/dsymbol.d]: OK. In that case I object to the code, it should not define `gramps`. Go with:

```d
if (!parent.toParent)
    // Insert a cool explanation here
    return null;
return parent.isSpeculative;
```

Comment (WalterBright) [inline src/dsymbol.d]: done

**Comment by andralex** [comment]
noice

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**